### ISSUE TITLE
Make Array.sort() and Array.sort_custom() return the array it sorted.

### ISSUE THREAD

**Issue by SlashScreen** [godotengine/godot-proposals#6877]
### Describe the project you are working on

No project in particular. I just like to chain commands together, and Godot is very close to being able to do that well with arrays

### Describe the problem or limitation you are having in your project

The process of list comprehension that GDScript 2 introduced is almost seamless, however, it is missing a few inlineable functions. The current method of sorting a new array is clunky:
```
var arr_1 = [1, 2, 3, 4]
var arr_2 = arr_1.filter(func(x:int): return x % 2 == 0)
arr_2.sort_custom(func(a:int, b:int): return a > b) # here we have to do an entire other step to sort the array
stringify_array(arr_2) # then we use the variable
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Having the function return the array it sorted would allow a more elegant list comprehension, without having to interrupt the code.
```
var arr_1 = [1, 2, 3, 4]
stringify_array(
    arr_1.filter(func(x:int): return x % 2 == 0)\
            .sort_custom(func(a:int, b:int): return a > b)
)
```

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

It would still modify the array directly, it would simply just return a reference to the array, instead of returning void.
```cpp
Array Array::sort_custom(const Callable &p_callable) {
	ERR_FAIL_COND_MSG(_p->read_only, "Array is in read-only state.");
	_p->array.sort_custom<CallableComparator, true>(p_callable);
	return _p; // Simply return the array afterwards.
}
```
This would complete the list comprehension workflow already started by gdscript 2, and make it feel less half-baked.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can, as detailed before; it's just clunky and feels half-baked.

### Is there a reason why this should be core and not an add-on in the asset library?

This alters the core functionality of the Array type.

**Comment by Calinou** [comment]
I believe returning a large array is expensive, so functions that operate on potentially large arrays are best kept in-place.

**Comment by SlashScreen** [comment]
Oh, does it not just return a pointer?

**Comment by mieldepoche** [comment]
maybe a `sorted` (and `sorted_custom` maybe? not a great name...) method should be added, that doesn't change the base array but return a sorted copy. That way you can use it in cases where perf won't matter and still have the sort option for the cases where it's a problem. I assume the typical case isn't sorting huge arrays on every frame, and I find being able to chain operations that way very convenient.

Although I don't understand either why returning an array could be expensive :thinking:

**Comment by KoBeWi** [comment]
Returning an Array is not expensive, because the method would simply return the same array reference.
It would be different with PackedArrays, as they have weird referencing rules.

**Comment by dalexeev** [comment]
> Although I don't understand either why returning an array could be expensive

`sort()` and `sort_custom()` sort an already existing array without creating a copy. The proposed method first creates a copy (deep or shallow?), which can be expensive in the case of a large array.

> maybe a `sorted` (and `sorted_custom` maybe?

I think this makes sense, but we should add a warning that `sort()` and `sort_custom()` are more efficient and preferred, unless you want to intentionally create a duplicate.

```gdscript
# Let's assume that array is the only owner of the reference.
# In this case, the immutable method is not recommended
# because it creates an extra duplicate:
array = array.sorted()

# You should use mutable method instead anyway:
array.sort()

# But if you want to intentionally create a copy and keep
# the original unsorted,then this code is ok:
var array2 = array.sorted()

# However, currently you can do this, it's just one more line,
# so I'm not sure if it's worth it:
var array2 = array.duplicate()
array2.sort()
```

**Comment by mieldepoche** [comment]
> I'm not sure if it's worth it

Well it would allow chaining operations, that was the initial point of the proposal:
```gdscript
var x := y.filter(...).map(...).sorted()
```

**Comment by Scony** [comment]
> maybe a `sorted` (and `sorted_custom` maybe? not a great name...) method should be added, that doesn't change the base array but return a sorted copy. That way you can use it in cases where perf won't matter and still have the sort option for the cases where it's a problem. I assume the typical case isn't sorting huge arrays on every frame, and I find being able to chain operations that way very convenient.
> 
> Although I don't understand either why returning an array could be expensive 🤔

Was just going to open proposal for that...

`.sorted()` and `sorted_custom()`, both having extra `deep=false` parameter should do the job just fine. Adding warnings makes no sense IMO. It just would be nice to have an explanation in the auto-generated API reference.

In practice, most people will use `sort()` anyway, but having `sorted()` for power-users would be really nice.

It resolves practical problems like e.g.
```
func render_the_data(data):  # data is some resource with fields
    # if data.items order matters, we'd need to create duplicate anyway
    for item in data.items.sorted():  # with sorted we can elegantly use reordered data.items w/o altering original array
        _render(item)
```

as well as a functional approach as mentioned by OP and @mieldepoche

**Comment by AThousandShips** [comment]
> Adding warnings makes no sense IMO

I'm assuming the warning is for the documentation, where it should be included

**Comment by Scony** [comment]
> > Adding warnings makes no sense IMO
> 
> I'm assuming the warning is for the documentation, where it should be included

Yeah, if it was meant for documentation then we all have full agreement on that.
